ipc_service: Automatically include source files for backends

eobalski opened the pull request:
The purpose of this change is to allow to enable more than one
backend at once.

The outcoming changes include:
- removal of choice option for IPC backends Kconfig
- removal of ZTEST backend
- removal necessity of explicitly specifying backend
- move IPC_SERVICE_NUM_ENDPOINTS_PER_INSTANCE option
  to Kconfig.rpmsg (+rename)
- automatical source file inclusion for chosen backend (now depends
  on DT configuration)
- clean up of samples and Kconfig files that were explicitly chosing
  STATIC_VRINGS as backend

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
Hi @kapi-no. I would like to kindly ask you to have a look here. I'm changing some things you've recently added. I think we don't need to explicitly select backends and this could be done as presented.
@e-rk if possible, could you verify if this suits you as well and if it affect your work in any significant way?
@carlocaione Those are the changes we discussed offline some time ago, please fell free to speak out.

emob-nordic commented:
> can you please split the BT changes from the general IPC ones in two separate commits?

IPC service won't work on 1st commit then. Are you sure about that? This change is pretty simple. I can do it but is there much need for that?

nordic-krch commented on samples/subsys/ipc/ipc_service/static_vrings/remote/prj.conf:
```diff
@@ -2,7 +2,6 @@ CONFIG_PRINTK=y
 CONFIG_HEAP_MEM_POOL_SIZE=4096
 
 CONFIG_IPC_SERVICE=y
-CONFIG_IPC_SERVICE_BACKEND_RPMSG=y
```
this could be split into another commit, since you are only changing the default in kconfig so keeping it here won't break any build, right?

carlocaione commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -1,46 +1,37 @@
-
 config IPC_SERVICE_RPMSG
 	bool "RPMsg support library"
+	depends on IPC_SERVICE_BACKEND_RPMSG
```
This doesn't actually depends on that backend. The backend is selecting this library, but the library can be selected by multiple backends. So, this is not needed.

carlocaione commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -1,46 +1,37 @@
 config IPC_SERVICE_BACKEND_RPMSG
 	bool "OpenAMP RPMSG backend with static VRINGs"
 	depends on MBOX
+	default $(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_IPC_OPENAMP_STATIC_VRINGS))
```
This confuses me. This change is implying that to select this backend we must change the DT. Without this change we are implying that we have to modify the defconfig instead. Now, do we have a strong preference for a way or the other?

emob-nordic replied:
I don't see any need to enforce the users to explicitly enable this option in configuration file. Afaik ipc-service was ment to be DTS configurable and this change actually does that. I just don't see any advantage of not making it this way. If the user adds `zephyr,ipc-openamp-static-vrings` compatible i believe he/she knows what he/she is doing.
This is more of a proposal rather than hard statement. I know you have objections about that, but i don't understand the justification of why not to do like so.

The only argument against such changes i heard in the past was: Kconfig is for software configuration and DT is for hardware configuration. If so - then configuring backend should never be DT configurable in the first place (imo). But we do accept that, so why not go one step further?

carlocaione replied:
> configuring backend should never be DT configurable in the first place (imo)

Right, but also Linux is going to have OpenAMP nodes in DT, so we are not the only ones :)

That said I don't have a strong opinion on this, so this is fine to me.

emob-nordic commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -1,46 +1,37 @@
 
 config IPC_SERVICE_STATIC_VRINGS_ALIGNMENT
 	int "VRINGs alignment"
+	depends on IPC_SERVICE_BACKEND_RPMSG
```
right.

Commit:
ipc_service: Automatically include source files for backends

The purpose of this change is to allow to enable more than one
backend at once by removing choice from ipc-service backend Kconfig
and depending backend Kconfig option on existing of correct compatible.

Overwriting IPC_SERVICE_BACKEND option in some places is removes
as no longer needed.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
ipc_service: samples: Remove redundant config options from samples

This change remove options that no longer needs to be set in
configuration file.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
Force pushed with changes + rebase due to conflicts.
- Removed depends on mentioned by @carlocaione and @nordic-krch  - see reviews above.
- Split removal of configuration options from samples to separate commit as it doesn't break anything.

nordic-krch commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -1,42 +1,32 @@
+
+config IPC_SERVICE_REG_BACKEND_PRIORITY
+	int "Initialization priority of modules registering IPC backend"
+	default 46
```
any chance to have static assert which would verify priority correctness? Something like: https://github.com/zephyrproject-rtos/zephyr/blob/93b0db781eafe9fbff24df6b4d31324c76c6941b/soc/arm/nordic_nrf/nrf53/sync_rtc.c#L310

emob-nordic replied:
Actually i just moved it, i don't know what was the intended relation. @carlocaione what should we put there?

carlocaione replied:
The only relation I can think of is with the MBOX driver that should be probed before the backend. Adding a `BUILD_ASSERT` for that is ok to me.

emob-nordic replied:
uhmmm, why is it `default 46` then? Default MBOX_INIT_PRIORITY is `Kernel:50`. For backend it would be `Kernel:46` now. Lower initialization priority means earlier initialization. Shouldn't it be > 50 then?
@carlocaione

carlocaione replied:
yeah, but that's not really import for the static vrings backend. That backend is initializing the mbox when opening the instance not on init. This change could be useful for other backends that are initializing the mbox during init.

carlocaione replied:
so, yeah, change the priority with respect to the mbox and add the assert even though that makes no difference for the current backend.

emob-nordic replied:
Shall we keep this option at all? I think there is no point in keeping that if any of the backends actually depends on `MBOX_INIT_PRIORITY`. Shall we use `CONFIG_KERNEL_INIT_PRIORITY_DEVICE` for both backends we have now?

emob-nordic replied:
I think there is no good reason to multiply options here if not needed. @nordic-krch  what do you think?

carlocaione replied:
please keep it. The point is that you can have some backend that is depending on some other driver that must be probed before the backend. So you can enforce a probe ordering only by acting on this symbol.

e-rk reviewed (dismissed):
No regressions observed with ieee802154 echo_client/server samples.

carlocaione commented on subsys/ipc/ipc_service/backends/Kconfig:
```diff
@@ -1,42 +1,32 @@
+	  The backend must be registered before the endpoint register.
 
 config IPC_SERVICE_BACKEND_RPMSG
 	bool "OpenAMP RPMSG backend with static VRINGs"
```
yes please. Some platforms would want to enable this using defconfig instead.

emob-nordic commented:
Rebased due to conflicts.

carlocaione commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -12,6 +12,11 @@
 #include <zephyr.h>
 #include <device.h>
 
+#if defined(CONFIG_MBOX_INIT_PRIORITY)
+BUILD_ASSERT(CONFIG_IPC_SERVICE_REG_BACKEND_PRIORITY > CONFIG_MBOX_INIT_PRIORITY,
+		"IPC service backend must be initialized after MBOX driver.");
+#endif
+
```
This should not be here because not necessarily the backend is using the MBOX. This is true for `static_vrings` but it's not a general solution.

emob-nordic replied:
Ah, yes. Sorry i forgot to move it.

emob-nordic replied:
Discussed offline. This assertion is not required for any backend right now. Im removing the last commit.
